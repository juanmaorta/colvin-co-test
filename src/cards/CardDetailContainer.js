import React, { Component } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { find } from 'lodash/collection'
import { withStyles } from '@material-ui/core/styles'

import { getWhiteCardsByParent, generateRandomEmptyCard } from '../util/cards'
import styles from '../styles'
import CardDetail from './components/CardDetail'
import {
  addCard,
  deleteCard,
  editCard
} from './actions'

export const searchCard = (cardList, id) => {
  const card = find(cardList, { id: id })
  if (!card) return

  card.whiteCards = getWhiteCardsByParent(cardList, id)

  return card
}

// maybe set card instead of cards?
const mapStateToProps = (state) => ({
  cards: state.cards
})

const mapDispatchToProps = (dispatch) => ({
  addCard: (parentId) => {
    return dispatch(addCard(generateRandomEmptyCard(parentId)))
  },
  deleteCard: (parentCard, whiteCard) => {
    return dispatch(deleteCard(parentCard, whiteCard))
  },
  editCard: () => {
    return dispatch(editCard())
  }
})

export class CardDetailContainer extends Component {
  constructor (props) {
    super(props)

    this.state = {
      card: searchCard(this.props.cards, this.props.match.params.cardId)
    }
  }

  static getDerivedStateFromProps (props, state) {
    const cardId = props.match.params.cardId
    const card = searchCard(props.cards, cardId)

    if (card && card.whiteCards.length !== state.card.whiteCards.length) {
      return {
        card
      }
    }
    return null
  }

  render () {
    const {
      classes,
      addCard,
      editCard,
      deleteCard
    } = this.props
    const { card } = this.state

    return (
      <div className={classes.heroUnit}>
        <div className={classes.heroContent}>
          <CardDetail
            card={card}
            addCard={() => addCard(card.id)}
            editCard={editCard}
            deleteCard={deleteCard}
          />
        </div>
      </div>
    )
  }
}

CardDetailContainer.propTypes = {
  classes: PropTypes.object.isRequired
}

export default withStyles(styles)(
  connect(mapStateToProps, mapDispatchToProps)(CardDetailContainer)
)
