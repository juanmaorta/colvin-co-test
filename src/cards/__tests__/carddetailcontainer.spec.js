/* global expect, describe, it, xit */
import React from 'react'
import { shallow } from 'enzyme'

import { CardDetailContainer, searchCard } from '../CardDetailContainer'
import * as stubs from './stubs'

describe('CardDetailContainer test suite', () => {
  describe('CardDetailContainer snapshots', () => {
    it('should match for CardDetailContainer component', () => {
      const match = {
        params: {
          cardId: 1
        }
      }

      const wrapper = shallow(
        <CardDetailContainer
          classes={stubs.classes}
          deleteCard={jest.fn}
          match={match}
        />
      )
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('searchCard', () => {
    it('should return undefined if card is not found', () => {
      expect(searchCard(stubs.cards, 12)).toBeUndefined()
    })
  })
})
