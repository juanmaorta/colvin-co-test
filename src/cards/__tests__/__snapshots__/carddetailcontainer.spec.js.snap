// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardDetailContainer test suite CardDetailContainer snapshots should match for CardDetailContainer component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CardDetailContainer
    classes={Object {}}
    deleteCard={[Function]}
    match={
        Object {
            "params": Object {
              "cardId": 1,
            },
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <WithStyles(CardDetail)
        addCard={[Function]}
        card={undefined}
        deleteCard={[Function]}
        editCard={undefined}
/>,
      "className": undefined,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "addCard": [Function],
        "card": undefined,
        "deleteCard": [Function],
        "editCard": undefined,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <WithStyles(CardDetail)
          addCard={[Function]}
          card={undefined}
          deleteCard={[Function]}
          editCard={undefined}
/>,
        "className": undefined,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "addCard": [Function],
          "card": undefined,
          "deleteCard": [Function],
          "editCard": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
